Add code cache flush mechanism

qwe661234 opened the pull request:
This modification introduces a code cache flush mechanism to handle the situation when the code cache becomes full. We flush the code cache by setting the offset to its initial value and clearing all the recording data for the code cache. Subsequently, we restart the T1C translation for the target chained-block.

Close: #365

jserv reviewed (commented):
<!-- github-benchmark-action-comment(start): Benchmarks Summary -->
# Benchmarks

<details>

| Benchmark suite | Current: 143a903047472cf85cdf94b40355bef18efacfc7 | Previous: 0d258e6d7d8259d86366811ac4d4cc3293c405d3 | Ratio |
|-|-|-|-|
| `Dhrystone` | `1750.62` Average DMIPS over 10 runs | `1744.6` Average DMIPS over 10 runs | `1.00` |
| `Coremark` | `1502.731` Average iterations/sec over 10 runs | `1496.516` Average iterations/sec over 10 runs | `1.00` |

</details>

This comment was automatically generated by [workflow](https://github.com/sysprog21/rv32emu/actions?query=workflow%3ABenchmark) using [github-action-benchmark](https://github.com/marketplace/actions/continuous-benchmark).
<!-- github-benchmark-action-comment(end): Benchmarks Summary -->

jserv commented on src/jit.c:
```diff
@@ -272,11 +272,11 @@ static inline void offset_map_insert(struct jit_state *state, int32_t target_pc)
     __builtin___clear_cache((char *) (addr), (char *) (addr) + (size));
 #endif
 
+static bool flush = false;
```
Rename to `should_flush` or something alike.

jserv commented on src/cache.c:
```diff
@@ -287,4 +287,14 @@ void cache_profile(const struct cache *cache,
 }
+
+void clear_cache_hot(const struct cache *cache, clear_func_t func)
+{
```
Add assertion on `cache` and `func`.

jserv commented on src/jit.c:
```diff
@@ -1684,28 +1703,29 @@ static void translate_chained_block(struct jit_state *state,
     state->n_jumps = 0;
-    uint32_t entry_loc = state->offset;
+    block->offset = state->offset;
     translate_chained_block(&(*state), rv, block);
```
Can you avoid using `&(*state)`?

jserv reviewed (changes requested):
Mention the preliminary explanation about the behavior of cache flushing.

qwe661234 committed:
Add code cache flush mechanism

This modification introduces a code cache flush mechanism to handle the
situation when the code cache becomes full. We flush the code cache by
setting the offset to its initial value and clearing all the recording
data for the code cache. Subsequently, we restart the T1C translation
for the target chained-block.

Close: #365
